chain/get_ancestors.go

```diff
@@ -85,57 +88,35 @@ func GetRecentAncestors(ctx context.Context, base types.TipSet, chainReader Read
 // CollectTipSetsOfHeightAtLeast collects all tipsets with a height greater
 // than or equal to minHeight from the input channel.  Precondition, the input
 // channel contains interfaces which may be tipsets or errors.
-func CollectTipSetsOfHeightAtLeast(ctx context.Context, ch <-chan interface{}, minHeight *types.BlockHeight) ([]types.TipSet, error) {
-	var ret []types.TipSet
+func CollectTipSetsOfHeightAtLeast(ctx context.Context, chainReader ReadStore, ts *types.TipSet, minHeight *types.BlockHeight) (ret []types.TipSet, err error) {
+	var h uint64
 	for {
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		case raw, more := <-ch:
-			if !more {
-				return ret, nil
-			}
-			switch raw := raw.(type) {
-			case error:
-				return nil, raw
-			case types.TipSet:
-				// Add tipset to ancestors.
-				h, err := raw.Height()
-				if err != nil {
-					return nil, err
-				}
-
-				// Check for termination.
-				if types.NewBlockHeight(h).LessThan(minHeight) {
-					return ret, nil
-				}
-				ret = append(ret, raw)
-			}
+		ret = append(ret, *ts)
+		ts, err = ts.GetNext(ctx, chainReader)
+		if ts == nil || err != nil {
+			return
+		}
+		// Add tipset to ancestors.
+		h, err = ts.Height()
```

Comment:
blocking: I believe we should return a nil slice in the case any error is thrown.  Because of the way you are doing named return variables a bunch of tipsets are being returned.

This belief comes from a memory of being told the same thing long ago on this project.  @phritz feel free to override this block if this isn't actually a big deal.

Reply:
If I'm right to block this then it also means the concise if statement on 96 will need to be broken into two or otherwise expanded a bit.

Reply:
same for 118 in `CollectAtMostNTipSets` below